### ISSUE TITLE
Excess number of queries generated with `get_subclass`

### ISSUE THREAD

**Issue by mjschultz** [jazzband/django-model-utils#274]
## Problem

After upgrading from 2.6.1 to 3.0.0 the number of queries generated by a `get_subclass` increases by (on average) half the number of subclasses there are.

## Environment

- Django Model Utils version: 3.0.0
- Django version: 1.10.7
- Python version: 3.4

## Code examples

Example `models.py` file
```
from django.db import models
from model_utils.managers import InheritanceManager


class BaseDemo(models.Model):
    my_field = models.IntegerField()
    objects = InheritanceManager()


class DemoA(BaseDemo):
    a_field = models.IntegerField()


class DemoB(BaseDemo):
    b_field = models.IntegerField()


class DemoC(BaseDemo):
    c_field = models.IntegerField()
```

Then in a Django shell I create an instance of each type and use `BaseDemo` to get a subclass:
```
from demo.models import *
DemoA.objects.get_or_create(my_field=1, a_field=2)
DemoB.objects.get_or_create(my_field=1, b_field=3)
DemoC.objects.get_or_create(my_field=1, c_field=4)
BaseDemo.objects.get_subclass(id=3)
```

In `django-model-utils==2.6.1` this would generate the following query and proper result:
```
SELECT "demo_basedemo"."id", "demo_basedemo"."my_field", "demo_demoa"."basedemo_ptr_id", "demo_demoa"."a_field", "demo_demob"."basedemo_ptr_id", "demo_demob"."b_field", "demo_democ"."basedemo_ptr_id", "demo_democ"."c_field" FROM "demo_basedemo" LEFT OUTER JOIN "demo_demoa" ON ("demo_basedemo"."id" = "demo_demoa"."basedemo_ptr_id") LEFT OUTER JOIN "demo_demob" ON ("demo_basedemo"."id" = "demo_demob"."basedemo_ptr_id") LEFT OUTER JOIN "demo_democ" ON ("demo_basedemo"."id" = "demo_democ"."basedemo_ptr_id") WHERE "demo_basedemo"."id" = 3; args=(3,)
```

with `django-model-utils==3.0.0` this generates the correct query, then an additional query for each sub-model-class:
```
SELECT "demo_basedemo"."id", "demo_basedemo"."my_field", "demo_demoa"."basedemo_ptr_id", "demo_demoa"."a_field", "demo_demob"."basedemo_ptr_id", "demo_demob"."b_field", "demo_democ"."basedemo_ptr_id", "demo_democ"."c_field" FROM "demo_basedemo" LEFT OUTER JOIN "demo_demoa" ON ("demo_basedemo"."id" = "demo_demoa"."basedemo_ptr_id") LEFT OUTER JOIN "demo_demob" ON ("demo_basedemo"."id" = "demo_demob"."basedemo_ptr_id") LEFT OUTER JOIN "demo_democ" ON ("demo_basedemo"."id" = "demo_democ"."basedemo_ptr_id") WHERE "demo_basedemo"."id" = 3; args=(3,)
SELECT "demo_basedemo"."id", "demo_basedemo"."my_field", "demo_demoa"."basedemo_ptr_id", "demo_demoa"."a_field" FROM "demo_demoa" INNER JOIN "demo_basedemo" ON ("demo_demoa"."basedemo_ptr_id" = "demo_basedemo"."id") WHERE "demo_demoa"."basedemo_ptr_id" = 3; args=(3,)
SELECT "demo_basedemo"."id", "demo_basedemo"."my_field", "demo_demob"."basedemo_ptr_id", "demo_demob"."b_field" FROM "demo_demob" INNER JOIN "demo_basedemo" ON ("demo_demob"."basedemo_ptr_id" = "demo_basedemo"."id") WHERE "demo_demob"."basedemo_ptr_id" = 3; args=(3,)
SELECT "demo_basedemo"."id", "demo_basedemo"."my_field", "demo_democ"."basedemo_ptr_id", "demo_democ"."c_field" FROM "demo_democ" INNER JOIN "demo_basedemo" ON ("demo_democ"."basedemo_ptr_id" = "demo_basedemo"."id") WHERE "demo_democ"."basedemo_ptr_id" = 3; args=(3,)
```

I say the number of calls increases by (on average) half the number of subclasses because they are sorted and it only generates queries up-to the point it finds the proper subclass. So if I did `BaseDemo.objects.get_subclass(id=1)`, it would generate the correct query and a single follow up query.

**Comment by sk1p** [comment]
I think I tracked this down. 8cb21aabbcba6aac389b2c073024fea08878126e notes:

> Django starting with 1.9 switched to using a class to provide an
iterator for the querymanager.  Between 1.9 and 1.10 changes slowly
stopped referencing that function and instead started calling
_iterator_class directly.

This may be true for Django 1.11, but on Django 1.9 and 1.10 ``QuerySet.iterator`` [is still called from](https://github.com/django/django/blob/1.10.7/django/db/models/query.py#L1087) ``QuerySet._fetch_all``:

    def _fetch_all(self):
        if self._result_cache is None:
            self._result_cache = list(self.iterator())
        if self._prefetch_related_lookups and not self._prefetch_done:
            self._prefetch_related_objects()

So, on Django 1.9 and 1.10, we have the following call chain: ``QuerySet._fetch_all`` → ``InheritanceQuerySetMixin.iterator()`` → ``QuerySet.iterator()`` → ``QuerySet._iterator_class()`` (that is ``InheritanceIterable``) returning back to ``InheritanceQuerySetMixin.iterator()``. Because the iteration logic was copied, we then [run all the ``_get_sub_obj_recurse`` logic again](https://github.com/jazzband/django-model-utils/blob/3.0.0/model_utils/managers.py#L110-L134), but on the selected subclass instances. Now, because the cached values from the `select_related(...)` call in `select_subclasses()` get lost when selecting the subclass instances, the database is hit again, _for ~ each item in the queryset!_

**Comment by smorokin** [comment]
@Natim: Is this issue closed with #389 (because of dropped support for older django versions)?

**Comment by Natim** [comment]
Yes good point.
